### PR TITLE
Drop Pipenv for Poetry

### PR DESCRIPTION
**Pull request by itsthejoker**
The toxicity of the pipenv maintainers and continuous instances of "oops, we broke it, guess you just have to wait until the next release" means that I don't want to associate our projects with their product. Moving `caroline` (and other projects, slowly) to Poetry.

### CODE REVIEW

**Comment by TheLonelyGhost** [comment]
Yep, that's one large reason I wasn't about to move us from setuptools-oriented installs to using Pipenv is because it still seems relatively unstable. I'm not sure yet how I feel about Poetry, but it did seem more stable.

**Comment by TheLonelyGhost** [comment]
Given the complete replacement of `setup.py` with Poetry's preferred TOML layout, what might that do for our dependency-links style dependencies in, for example, tor? Have we decided on publishing Caroline to pypi, or is that still up in the air?

I have to say I do like some of the mechanisms Poetry gives us for publishing, where I used to just use [`twine`](https://pypi.org/project/twine/) instead. I also like that it will defer to already-setup virtualenvs in my case where I have a [direnv](https://direnv.net)-based python development workflow.

All the flair of progress bars in the CLI makes me wonder how scriptable poetry actually is in the standard unix-y way. Can I pipe to it? Use grep? It seems similar to how `yarn` handles things as compared to `npm`, which has been proven to include some trade-offs. I wonder what those might be for python... 🤔

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -4,7 +4,3 @@ repos:
     hooks:
     - id: black
       language_version: python3.7
--   repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v1.2.3
-    hooks:
-    - id: flake8
```
Comment (TheLonelyGhost) [inline .pre-commit-config.yaml:10]: 👎 :trollface:

File: .travis.yml [hunk 2]
```diff
@@ -19,9 +19,12 @@ matrix:
...
 install:
-  - pip install --process-dependency-links .[dev]
-  - pip install codacy-coverage
-  - pip install pytest-cov
+  - pip install pip -U
+  - wget https://raw.githubusercontent.com/sdispater/poetry/master/get-poetry.py
+  - python get-poetry.py --preview -y
+  - source $HOME/.poetry/env
+  - poetry install -v
+  - poetry add codacy-coverage pytest-cov
```
Comment (TheLonelyGhost) [inline .travis.yml]: Should we instead maintain a CI-specific set of dependencies? We can install them with Poetry's [support for "extras"](https://poetry.eustace.io/docs/pyproject/#extras). Example:

```toml
[tool.poetry.extras]
ci = ["code-coverage", "pytest-cov"]
```

```suggestion
  - poetry install -v -E ci
  - poetry add codacy-coverage pytest-cov
```

File: poetry.lock [hunk 3]
```diff
@@ -0,0 +1,542 @@
+[[package]]
```
Comment (TheLonelyGhost) [inline poetry.lock:1]: Do we need the `poetry.lock` file if this is [intended to be a library](https://poetry.eustace.io/docs/basic-usage/#commit-your-poetrylock-file-to-version-control)?

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+jsonschema = "^2.6"
+elasticsearch = "^6.3"
+addict = "^2.2"
+
+[tool.poetry.dev-dependencies]
+pytest = "^3.10"
+pytest-cov = "^2.6"
+black = "^18.3-alpha.0"
+ipdb = "^0.11.0"
+pre-commit = "^1.12"
```
Comment (TheLonelyGhost) [inline pyproject.toml:36]: ❤️

File: .gitignore [hunk 5]
```diff
@@ -102,3 +102,6 @@ venv.bak/
 
 # mypy
 .mypy_cache/
+
+.idea/
+asdf.py
```
Comment (TheLonelyGhost) [inline .gitignore:107]: Why do we need to add this to the gitignore?

Comment (itsthejoker) [reply .gitignore:107]: We can rename that to something more reasonable; that's my file that has a complete mockup application for manually testing the library.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (itsthejoker) [reply .pre-commit-config.yaml:10]: Black forces a lot of the things that flake8 complains about, and if there's an illegal line then black should explode anyway. I can re-add it, it just seemed extraneous when we're already forcing a code style.

File: .travis.yml (hunk 2 above)
Comment (itsthejoker) [reply .travis.yml]: That's a great idea and I'll try to look into that tonight.

**Comment by itsthejoker** [comment]
@TheLonelyGhost I've been using poetry pretty solidly over the last week. I've only run into one major issue, and I just fixed it myself in my shell: https://github.com/sdispater/poetry/issues/497

That being said, `caroline` 0.4.1 is on pypi already and has been for a little while. I don't have any issues installing it either with pip, pipenv, or poetry. I can confirm that `poetry build` works flawlessly, but I haven't published the new version because we're still discussing in this PR 😄 

As for `tor`, installing directly from github does work, and we can add that into the `pyproject.toml` file.

```sh
 ~/g/caroline   poetry   poetry add tor_core --git https://github.com/grafeasgroup/tor_core

Updating dependencies
Resolving dependencies... (7.8s)


Package operations: 22 installs, 0 updates, 0 removals

Writing lock file

  - Installing certifi (2018.10.15)
  - Installing chardet (3.0.4)
  - Installing idna (2.7)
  - Installing jaraco.functools (1.20)
  - Installing pytz (2018.7)
  - Installing backports.functools-lru-cache (1.5)
  - Installing requests (2.20.1)
  - Installing tempora (1.14)
  - Installing cheroot (6.5.2)
  - Installing portend (2.3)
  - Installing prawcore (0.11.0)
  - Installing update-checker (0.16)
  - Installing webob (1.8.3)
  - Installing websocket-client (0.54.0)
  - Installing zc.lockfile (1.3.0)
  - Installing bugsnag (3.4.3)
  - Installing cherrypy (18.0.1)
  - Installing praw (5.0.1)
  - Installing raven (6.9.0)
  - Installing sh (1.12.14)
  - Installing slackclient (1.3.0)
  - Installing tor-core (0.5.0 50f3b1e)

 ~/g/caroline   poetry    poetry run python                                  44.6s  Fri Nov  9 19:19:39 2018
Python 3.7.0 (default, Oct 28 2018, 22:12:20)
[Clang 10.0.0 (clang-1000.10.44.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from tor_core import initialize
>>>
```
More in https://github.com/sdispater/poetry/issues/313.

File: poetry.lock (hunk 3 above)
Comment (itsthejoker) [reply poetry.lock:1]: Yes, I'd like to keep it for local development. It doesn't really matter in the grand scheme of things but it's helpful if anyone else wants to work on this. Also for CI.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (TheLonelyGhost) [reply .pre-commit-config.yaml:10]: Yep, I just like flake8. It's a design choice, but I'm on board with it. :smile:

**Comment by itsthejoker** [comment]
Approved by @TheLonelyGhost in slack, merging and publishing!

**Pull request merged**